### DISCUSSION TITLE
Convert molecule fragments to SMARTS and keep atomic features

### DISCUSSION THREAD

**Discussion by Minys233** [rdkit/rdkit#7693] [Q&A] [answered]
I'm breaking bonds with `Chem.FragmentOnBonds` and then use `Chem.GetMolFrags` to get some molecule fragments for later template matching. I found the atomic features are not included in the SMARTS if I use `Chem.MolToSmarts` to the fragments, this will make the template matching not accurate.

```python
from rdkit import Chem
mol = Chem.MolFromSmiles('CC')
frags = Chem.FragmentOnBonds(mol, [0], addDummies=False)
frag = Chem.GetMolFrags(frags, asMols=True, sanitizeFrags=True)[0]
print(Chem.MolToSmarts(frag))
another_mol = Chem.MolFromSmiles('CC(=O)OC1=CC=CC=C1C(=O)O')
print(another_mol.GetSubstructMatches(frag))

## Output
[#6]
((0,), (1,), (4,), (5,), (6,), (7,), (8,), (9,), (10,))
```
<img width="200" alt="image" src="https://github.com/user-attachments/assets/498144a6-a79f-4893-af94-69738f03a53a">

In this example, the `frag` is a methyl group, and the corresponding SMARTS should be `"[C&X4&H3]"`, which should only match one terminal carbon atom (figure below), rather than all carbon atoms (figure above).

<img width="200" alt="image" src="https://github.com/user-attachments/assets/eae7d87c-e9bf-4d04-bbce-87c4840496f9">

**So my question is: how can I controlable include atomic features in SMARTS for arbitrary Mol objects initialized from SMILES?**

In addition to search in issues & discussions, also asked GPT-4o and Claude-3.5-Sonnet and they just said to manually add such features, but this strategy cannot adapt to complex patterns with branches and rings.

Thank you for your time and patience.

**Comment by DavidACosgrove** [comment, marked as answer]
Hi @Minys233
Surprisingly, at least to my cynical old mind, the AI programs have given you a correct answer, albeit not a very helpful one.  This is my solution, which is not necessarily optimal or elegant.
```
# Set a prop on the atoms in the mol to be fragmented so we can identify them later
for atom in mol.GetAtoms():
    atom.SetProp("idx", f"{atom.GetIdx()}")

# fragment the molecule and take the second fragment
frags = Chem.FragmentOnBonds(mol, [0], addDummies=False)
frag = Chem.GetMolFrags(frags, asMols=True, sanitizeFrags=True)[1]
# convert the frag into a QueryMol by replacing each atom with a QueryAtom,
# and transfer the properties of interest from the original molecule as
# query parameters to the new query atom.  The atom properties are all
# copied over from the original
for atom in frag.GetAtoms():
    # get the index of the atom in the original molecule
    print(f'frag atom index : {atom.GetProp("idx")}')
    atidx = int(atom.GetProp("idx"))
    # make a new QueryAtom in the frag
    qatom = rdqueries.ReplaceAtomWithQueryAtom(frag, atom)
    # get the corresponding atom from the original mol
    oldatom = mol.GetAtomWithIdx(atidx)
    print(f"num hs : {oldatom.GetTotalNumHs()}  degree : {oldatom.GetDegree()}")
    # Add queries for explicit degree and hydrogen count to the query
    qa = rdqueries.ExplicitDegreeEqualsQueryAtom(oldatom.GetDegree())
    qatom.ExpandQuery(qa, Chem.CompositeQueryType.COMPOSITE_AND)
    qa = rdqueries.HCountEqualsQueryAtom(oldatom.GetTotalNumHs())
    qatom.ExpandQuery(qa, Chem.CompositeQueryType.COMPOSITE_AND)

print(f"SMARTS of query : {Chem.MolToSmarts(frag)}")
```
This produces the output
<img width="455" alt="image" src="https://github.com/user-attachments/assets/8d362913-5794-4968-ac19-294d6785411c">

You can set a number of different properties on the query atom to check for things like ring membership, bridgeheads etc.  

HTH,
Dave

**Reply by Minys233** [reply]
Thank you so much for sharing insight and I will try this idea very soon. I probably understand what you're doing, replacing atoms with `QueryAtom` and then convert them to SMARTS. So sorry, I didn't know about `QueryAtom` and `QueryMol` classes before and there aren't many descriptions on the internet. Maybe that's the key to the AI not being able to give a good answer :P Thanks again!

Best wishes,
Yaosen

**Comment by bp-kelley** [comment]
I think there might be a better solution.    In the RDKit molecules and queries share a lot of machinery.

There is a function, AdjustQueryProperties that can add the degree queries for you.  Then we can simply add and merge hydrogens to get a suitable query:

```
>>> mol = Chem.MolFromSmiles("CC")
# we need to add explicit degree BEFORE adding hydrogens
>>> params = Chem.AdjustQueryParameters()
>>> params.adjustDegree = True
>>> params.adjustDegreeFlags = Chem.ADJUST_IGNOREDUMMIES
>>> molq = Chem.AdjustQueryProperties(mol,params)
# now add and merge query hydrogens
>>> Chem.MolToSmarts(Chem.MergeQueryHs(Chem.AddHs(molq)))
'[#6&D1&!H0&!H1&!H2]-[#6&D1&!H0&!H1&!H2]'
```

Note that "D" is explicit degree, but since smiles are hydrogen suppressed, this is why we need to add it first.

Also because smarts describes a SubstructureMatch, it is allowed to match things with more than N hydrogen atoms, but I think that's ok for your purposes.
